node/src/components/block_validator.rs

```diff
@@ -270,43 +251,34 @@ where
-                            // caller.
-                            effects.extend(responder.respond(true).ignore());
-                        } else {
-                            // We register ourselves as someone interested in the ultimate
-                            // validation result.
-                            entry.get_mut().responders.push(responder);
-                            // And add an alternative source of data.
-                            entry.get_mut().add_source(sender);
-                        }
-                    }
-                    Entry::Vacant(entry) => {
-                        // Our entry is vacant - create an entry to track the state.
-                        let in_flight = &mut self.in_flight;
-                        effects.extend(entry.key().deploys_and_transfers_iter().flat_map(
-                            |(dt_hash, _)| {
-                                // For every request, increase the number of in-flight...
-                                in_flight.inc(&dt_hash.into());
-                                // ...then request it.
-                                fetch_deploy(effect_builder, dt_hash, sender.clone())
-                            },
-                        ));
-                        let block_timestamp = entry.key().timestamp();
-                        let deploy_config = self.chainspec.deploy_config;
-                        entry.insert(BlockValidationState {
-                            appendable_block: AppendableBlock::new(deploy_config, block_timestamp),
-                            missing_deploys: block_deploys,
-                            responders: smallvec![responder],
-                            sources: VecDeque::new(), /* This is empty b/c we create the first
-                                                       * request using `sender`. */
-                        });
-                    }
+                let block_timestamp = block.timestamp();
+                let state = self
+                    .validation_states
+                    .entry(block)
+                    .or_insert(BlockValidationState {
+                        appendable_block: AppendableBlock::new(
+                            self.chainspec.deploy_config,
+                            block_timestamp,
+                        ),
+                        missing_deploys: block_deploys.clone(),
+                        responders: smallvec![],
+                    });
+
+                if state.missing_deploys.is_empty() {
+                    // Block has already been validated successfully, early return to caller.
+                    return responder.respond(true).ignore();
                 }
+
+                // We register ourselves as someone interested in the ultimate validation result.
+                state.responders.push(responder);
+
+                let in_flight = &mut self.in_flight;
```

Comment:
nit: would have liked to not take this extra reference and just use `self.in_flight` to increment.